File: HW_1/hw1_color.py
import cv2, skimage.data
import numpy as np
import matplotlib.pyplot as plt

# Load Image
img = skimage.data.astronaut()

# 1) RGB, Python
R = img[:, :, 0]
G = img[:, :, 1]
B = img[:, :, 2]
cv2.imwrite('./fig/RGB_astronaut.png', cv2.hconcat([R, G, B]))

# 2) YCbCr, Python
YCbCr = np.array([[ 77,  150,  29],
                  [-43,  -84, 127],
                  [127, -106, -21]]) / 256.
img_YCbCr = img.dot(YCbCr.T)
Y, Cb, Cr = np.split(img_YCbCr, 3, axis=2)
Cb = np.clip(Cb+128, 0, 255)
Cr = np.clip(Cr+128, 0, 255)
#Y = np.clip(Y, 0, 255).astype(np.uint8)
#Cb = np.clip(Cb, 0, 255).astype(np.uint8)
#Cr = np.clip(Cr, 0, 255).astype(np.uint8)
cv2.imwrite('./fig/YCbCr_astronaut.png', cv2.hconcat([Y, Cb, Cr])) 

# 3) HSI, Python
eps = 1e-8
R, G, B = R/255, G/255, B/255
I = (R + G + B) / 3
S = 1 - np.minimum(np.minimum(R, G), B) / (I+eps)
H = 1 / (2*np.pi) * np.arccos((2*R-G-B)/(2*np.sqrt((R-G)**2+(R-B)*(G-B)+eps)))
H[B>G] = 1 - H[B>G]
H[I==0] = 0
cv2.imwrite('./fig/HSI_astronaut.png', cv2.hconcat([H*255, S*255, I*255]))

# 4) Modifying image
S += 40/255
S[S>=1] = 1
I += 16/255
I[I>=1] = 1
hsi = np.zeros(img.shape)
for i in range(img.shape[0]):
    for j in range(img.shape[1]):
        if H[i, j]>=0 and H[i, j]<1/3:
            b = I[i, j] * (1 - S[i, j])
            r = I[i, j] * (1 + S[i, j]*np.cos(H[i, j]*2*np.pi)/np.cos(np.pi/3 - H[i, j]*2*np.pi))
            g = 3*I[i, j] - (r + b)
        elif H[i, j]>=1/3 and H[i, j]<2/3:
            r = I[i, j] * (1 - S[i, j])
            g = I[i, j] * (1 + S[i, j]*np.cos(H[i, j]*2*np.pi-2*np.pi/3)/np.cos(np.pi - H[i, j]*2*np.pi))
            b = 3*I[i, j] - (r + g)
        else:
            g = I[i, j] * (1 - S[i, j])
            b = I[i, j] * (1 + S[i, j]*np.cos(H[i, j]*2*np.pi-4*np.pi/3)/np.cos(5*np.pi/3-H[i, j]*2*np.pi))
            r = 3*I[i, j] - (g + b)
        hsi[i, j, 0], hsi[i, j, 1], hsi[i, j, 2] = r, g, b
hsi *= 255
cv2.imwrite('./fig/HSI2RGB_astronaut.png', hsi[:, :, ::-1])

File: HW_1/Submit/hw1_color.m
clear all;

%img = imread('peppers.png');
img = imread('./fig/cb.png');

% 1) RGB, Matlab
R = img(:, :, 1);
G = img(:, :, 2);
B = img(:, :, 3);
%imwrite([R, G, B], './fig/RGB_peppers.png');
imwrite([R, G, B], './fig/RGB_cb.png');

% 2) YCbCr, Matlab
YCbCr = [ 77  150   29; 
         -43  -84  127; 
         127 -106  -21;] / 256.;
img_YCbCr = reshape(double(img)/255, [], 3) * YCbCr';
img_YCbCr = reshape(img_YCbCr, size(img));
Y = img_YCbCr(:, :, 1);
Cb = min(img_YCbCr(:, :, 2) + 0.5, 1);
Cr = min(img_YCbCr(:, :, 3) + 0.5, 1);
%imwrite([Y, Cb, Cr], './fig/YCbCr_peppers.png');
imwrite([Y, Cb, Cr], './fig/YCbCr_cb.png');

% 3) HSI, Matlab
R = double(R) / 255;
G = double(G) / 255;
B = double(B) / 255;
I = (R + G + B) / 3;
S = 1 - min(min(R, G), B) ./ (I+eps);
H = 1 / (2*pi) * acos((2*R-G-B)./(2*sqrt((R-G).^2+(R-B).*(G-B))+eps));
H(B>G) = 1 - H(B>G);
H(I==0) = 0;
%imwrite([H, S, I], './fig/HSI_peppers.png');
imwrite([H, S, I], './fig/HSI_cb.png');

% 4) Modifying Image
S = S+40/255;
S(S>=1) = 1;
I = I+16/255;
I(I>=1) = 1;
hsi = zeros(size(R, 1), size(R, 2), 3);
for i = 1:size(R, 1)
    for j = 1:size(R, 2)
        if H(i, j)>=0 && H(i, j)<1/3
            b = I(i, j) * (1 - S(i, j));
            r = I(i, j) * (1 + S(i, j)*cos(H(i, j)*2*pi)/cos(pi/3 - H(i, j)*2*pi));
            g = 3*I(i, j) - (r + b);
        elseif H(i, j)>=1/3 && H(i, j)<2/3
            r = I(i, j) * (1 - S(i, j));
            g = I(i, j) * (1 + S(i, j)*cos(H(i, j)*2*pi - 2*pi/3)/cos(pi - H(i, j)*2*pi));
            b = 3*I(i, j) - (r + g);
        else
            g = I(i, j) * (1 - S(i, j));
            b = I(i, j) * (1 + S(i, j)*cos(H(i, j)*2*pi - 4*pi/3)/cos(5*pi/3 - H(i, j)*2*pi));
            r = 3*I(i, j) - (g + b);
        end
        hsi(i, j, 1) = r;
        hsi(i, j, 2) = g;
        hsi(i, j, 3) = b;
    end
end
%imwrite(hsi, './fig/HSI2RGB_peppers.png');
imwrite(hsi, './fig/HSI2RGB_cb.png');